ngraph/frontend/onnx_editor/include/onnx_editor/edge_mapper.hpp

```diff
@@ -0,0 +1,95 @@
+            ///
+            /// \note If state of graph_proto will be changed, the information
+            ///       from edge mapper is outdated.
+            ///
+            /// \param graph_proto Reference to GraphProto object.
+            EdgeMapper(const ONNX_NAMESPACE::GraphProto& graph_proto);
+
+            /// \brief Returns the InputEdge based on a node (node name or output name)
+            ///        and an input (input name or input index).
+            ///
+            /// \note  The node name can be ambiguous (many nodes can have the same name).
+            ///        In such a case the algorthim try to match the given node name
+            ///        with the input name (given input index is not enough).
+            ///        If after such operation a found edge is unique, it is returned.
+            ///        If InputEdge cannot be determined based on given params the ngraph_error
+            ///        exception is thrown.
+            ///
+            /// \param node A node helper structure created based on a node name
+            ///             or a node output name.
+            ///
+            /// \param input A input helper structure created based on a input name
+            ///              or a input index.
+            InputEdge to_input_edge(const Node& node, const Input& input) const;
+
+            /// \brief Returns the OutputEdge based on a node (node name or output name)
+            ///        and an output (output name or output index).
+            ///
+            /// \note  The node name can be ambiguous (many nodes can have the same name).
+            ///        In such a case the algorthim try to match the given node name
+            ///        with the output name (given output index is not enough).
+            ///        If after such operation a found edge is unique, it is returned.
+            ///        If OutputEdge cannot be determined based on given params the ngraph_error
+            ///        exception is thrown.
+            ///
+            /// \param node A node helper structure created based on a node name
+            ///             or a node output name.
+            ///
+            /// \param output A output helper structure created based on a output name
+            ///               or a output index.
+            OutputEdge to_output_edge(const Node& node, const Output& output) const;
+
+        private:
+            std::vector<int> find_node_indexes(const std::string& node_name,
+                                               const std::string& output_name) const;
+            std::string get_node_input_name(int node_index, int input_index) const;
+            std::string get_node_output_name(int node_index, int output_index) const;
+
+            std::vector<std::vector<std::string>> m_node_inputs;
+            std::vector<std::vector<std::string>> m_node_outputs;
+            std::multimap<std::string, int> m_node_name_to_index;
```

Comment:
I wonder if there is a worthy possibility to avoid caching all of these names.

Reply:
The current approach has the following advantages IMO:
- lifetime of `EdgeMapper` is independent to `ONNXEditor`
- time-consuming operations are performed once in the constructor
- once instance of `EdgeMapper` return always the same results
- `EdgeMapper` instance represents some state (snapshot) in ONNXEditor lifetime